File: 1337-design-skiplist/design-skiplist.py
import random

class Node:
    def __init__(self, val: int, level: int):
        self.val = val
        self.forward = [None] * level


class Skiplist:
    MAX_LEVEL = 16
    P = 0.5

    def __init__(self):
        self.head = Node(-1, self.MAX_LEVEL)
        self.level = 1

    def _random_level(self) -> int:
        lvl = 1
        while lvl < self.MAX_LEVEL and random.random() < self.P:
            lvl += 1
        return lvl

    def search(self, target: int) -> bool:
        curr = self.head
        for i in range(self.level - 1, -1, -1):
            while curr.forward[i] and curr.forward[i].val < target:
                curr = curr.forward[i]
        curr = curr.forward[0]
        return curr is not None and curr.val == target

    def add(self, num: int) -> None:
        update = [None] * self.MAX_LEVEL
        curr = self.head

        for i in range(self.level - 1, -1, -1):
            while curr.forward[i] and curr.forward[i].val < num:
                curr = curr.forward[i]
            update[i] = curr

        lvl = self._random_level()
        if lvl > self.level:
            for i in range(self.level, lvl):
                update[i] = self.head
            self.level = lvl

        node = Node(num, lvl)
        for i in range(lvl):
            node.forward[i] = update[i].forward[i]
            update[i].forward[i] = node

    def erase(self, num: int) -> bool:
        update = [None] * self.MAX_LEVEL
        curr = self.head

        for i in range(self.level - 1, -1, -1):
            while curr.forward[i] and curr.forward[i].val < num:
                curr = curr.forward[i]
            update[i] = curr

        curr = curr.forward[0]
        if curr is None or curr.val != num:
            return False

        for i in range(self.level):
            if update[i].forward[i] != curr:
                break
            update[i].forward[i] = curr.forward[i]

        while self.level > 1 and self.head.forward[self.level - 1] is None:
            self.level -= 1

        return True

# Your Skiplist object will be instantiated and called as such:
# obj = Skiplist()
# param_1 = obj.search(target)
# obj.add(num)
# param_3 = obj.erase(num)

File: 1337-design-skiplist/design-skiplist.cpp
#include <random>
#include <vector>

class Skiplist {
private:
    static constexpr int MAX_LEVEL = 16;
    struct Node {
        int value;
        std::vector<Node*> next;
        Node (int value, int level)
            : value(value), next(level, nullptr) {}
    };
    Node* head;
    std::mt19937 generator;
    std::bernoulli_distribution coinFlip;

    int randomLevel() {
        int level = 1;
        while (level < MAX_LEVEL && coinFlip(generator)) {
            ++level;
        }
        return level;
    }
public:
    Skiplist()
        : head(new Node(0, MAX_LEVEL)),
          generator(std::random_device{}()),
          coinFlip(0.5) {}
    ~Skiplist() {
        Node* current = head;
        while (current != nullptr) {
            Node*next = current->next[0];
            delete current;
            current = next;
        }
    }
    
    bool search(int target) {
        Node* current = head;
        for (int level = MAX_LEVEL - 1; level >= 0; --level) {
            while (current->next[level] != nullptr && current->next[level]->value < target) {
                    current = current->next[level];
            }
        }
        current = current->next[0];
        return current != nullptr && current->value == target;
    }
    
    void add(int num) {
        std::vector<Node*> predecessors(MAX_LEVEL, nullptr);
        Node* current = head;
        for (int level = MAX_LEVEL - 1; level >= 0; --level) {
            while (current->next[level] != nullptr && current->next[level]->value < num) {
                current = current->next[level];
            }
            predecessors[level] = current;
        }
        int nodeLevel = randomLevel();
        Node* newNode = new Node(num, nodeLevel);
        for (int level = 0; level < nodeLevel; ++level) {
            newNode->next[level] = predecessors[level]->next[level];
            predecessors[level]->next[level] = newNode;
        }
    }
    
    bool erase(int num) {
        std::vector<Node*> predecessors(MAX_LEVEL, nullptr);
        Node* current = head;
        for (int level = MAX_LEVEL - 1; level >= 0; --level) {
            while (current->next[level] != nullptr && current->next[level]->value < num) {
                current = current->next[level];
            }
            predecessors[level] = current;
        }
        Node* target = predecessors[0]->next[0];
        if (target == nullptr || target->value != num) {
            return false;
        }
        for (int level = 0; level < MAX_LEVEL; ++level) {
            if (predecessors[level]->next[level] == target) {
                predecessors[level]->next[level] = target->next[level];
            }
        }
        delete target;
        return true;
    }
};

/**
 * Your Skiplist object will be instantiated and called as such:
 * Skiplist* obj = new Skiplist();
 * bool param_1 = obj->search(target);
 * obj->add(num);
 * bool param_3 = obj->erase(num);
 */